File: chapters/monte_carlo/code/swift/monte_carlo.swift
//Double Extension from YannickSteph on StackOverflow: https://stackoverflow.com/questions/25050309/swift-random-float-between-0-and-1

import Foundation



public extension Double {

    public static var random: Double {
        
        return Double(arc4random()) / 0xFFFFFFFF
    }

    public static func random(min: Double, max: Double) -> Double {

        return Double.random * (max - min) + min
    }
}


func isInCircle(x: Double, y: Double, radius: Double) -> Bool {

    return (x*x) + (y*y) < radius*radius
}


func monteCarlo(n: Int, radius: Double) -> Double {

    var piCount = 0
    var randX: Double
    var randY: Double

    for _ in 0...n {
        randX = Double.random(min: 0, max: radius)
        randY = Double.random(min: 0, max: radius)

        if(isInCircle(x: randX, y: randY, radius: radius)) {
            piCount += 1
        }
    }
    
    let piEstimate = Double(4 * piCount)/(Double(n))
    print("Percent error is: \(100*(Double.pi - piEstimate)/Double.pi)%")

    return piEstimate
}


func main() {
    print("Pi estimate is: ", monteCarlo(n: 10000, radius: 50))
}


main()

File: chapters/monte_carlo/code/csharp/MonteCarlo.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace MonteCarlo
{
    class MonteCarlo
    {
        private static Random random = new Random(47);
        private readonly int radius = 1;


        public MonteCarlo(int radius)
        {
            this.radius = radius;
        }


        public double Run(int samples)
        {
            int piCount = 0;

            for (int i = 0; i < samples; i++)
            {
                var randomX = random.NextDouble() * radius;
                var randomY = random.NextDouble() * radius;

                if (IsInCircle(randomX, randomY)) piCount++;
            }

            return 4 * piCount / (double)samples;
        }


        private bool IsInCircle(double x, double y) => x * x + y * y < radius * radius;
    }
}
